File: DSA/2. Troubling Traversals/Part 5 (Hard)/Python/judge.py
from collections import defaultdict, deque
from solution import Solution

class Judge:
    def run_tests(self):
        test_cases_paths = [
            [
                [2, 1],
                [0, 1],
                [1, 3],
                [5, 3],
                [2, 4],
                [3, 6]
            ],
            [
                [0,1],
                [1,3],
                [2,3],
                [4,0],
                [4,5]
            ]
        ];

        test_cases_n = [
            7,
            6
        ];

        for test_case_paths, test_case_n in zip(test_cases_paths, test_cases_n):
            self.test(test_case_paths, test_case_n)

    def test(self, test_case_paths, test_case_n):
        solution = Solution()
        actual = solution.redirect_paths(test_case_paths, test_case_n)
        expected = self.test_case_solution_output(test_case_paths, test_case_n)
        passed = actual == expected
        print(f"Nodes: {test_case_n}")
        print(f"Paths: {test_case_paths}")
        print(f"Expected: {expected}")
        print(f"Actual: {actual}")
        print("Test " + ("PASSED" if passed else "FAILED"))
        print("-----------")

    def test_case_solution_output(self, paths, n):
        graph = defaultdict(list)
        reversed_edges = set()

        for from_node, to_node in paths:
            graph[from_node].append(to_node)
            graph[to_node].append(from_node)
            reversed_edges.add((from_node, to_node))

        swaps = 0
        visited = set()
        stack = deque([0])

        while stack:
            node = stack.pop()
            visited.add(node)

            for neighbor in graph[node]:
                if neighbor not in visited:
                    if (node, neighbor) in reversed_edges:
                        swaps += 1
                    stack.append(neighbor)

        return swaps
        

File: DSA/2. Troubling Traversals/Part 2 (Easy)/JavaScript/judge.js
const Solution = require('./solution');

class Judge {
    runTests() {
        const testCasesT = [
            [1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1],
            [1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1],
        ];

        const testCasesB = [
            5,
            3
        ];

        for (let i = 0; i < testCasesT.length; i++) {
            test(testCasesT[i], testCasesB[i], testCaseSolutionOutput(testCasesT[i], testCasesB[i]));
        }
    }

    test(t, b, expected) {
        const solution = new Solution();
        const actual = solution.buildBridges(t, b);
        const passed = JSON.stringify(actual) === JSON.stringify(expected);

        console.log(`Terrain: ${JSON.stringify(t)}`);
        console.log(`Bridges: ${b}`);
        console.log(`Expected: ${JSON.stringify(expected)}`);
        console.log(`Actual: ${JSON.stringify(actual)}`);
        console.log(passed ? "Test PASSED" : "Test FAILED");
        console.log("-----------");
    }

    testCaseSolutionOutput(t, b) {
        const n = t.length;
        let bridgesUsed = 0;
        for (let i = 0; i < n; i++) {
            if (t[i] == 0) {
                let start = i;
                while (i < n && t[i] == 0) {
                    i++;
                }
                const length = i - start;
                if (length > 1) {
                    if (bridgesUsed + (length - 1) > b) {
                        return [];
                    }
                    for (let j = start; j < i - 1; j++) {
                        t[j] = 2;
                        bridgesUsed++;
                    }
                }
            }
        }
        return t;
    }
}

module.exports = Judge;